File: backend/app/schemas/company.py
"""Schemas for company tenancy and employer team management."""

from datetime import datetime

from typing import Any

from pydantic import BaseModel, ConfigDict, EmailStr, Field, model_validator

from app.models.company import (
    InvitationDeliveryStatus,
    InvitationStatus,
    MembershipRole,
    MembershipStatus,
)


class CompanyRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    description: str | None = None
    logo_url: str | None = None
    is_active: bool


class DepartmentCreate(BaseModel):
    name: str = Field(min_length=2, max_length=150)
    description: str | None = Field(default=None, max_length=1000)


class DepartmentUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=2, max_length=150)
    description: str | None = Field(default=None, max_length=1000)
    is_active: bool | None = None


class DepartmentRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    company_id: int
    name: str
    description: str | None = None
    is_active: bool


class MembershipRead(BaseModel):
    id: int
    company_id: int
    user_id: int
    full_name: str
    email: str
    avatar_url: str | None = None
    member_role: MembershipRole
    department_id: int | None = None
    department_name: str | None = None
    status: MembershipStatus
    is_owner: bool
    joined_at: datetime
    updated_at: datetime


class MembershipUpdate(BaseModel):
    member_role: MembershipRole | None = None
    department_id: int | None = None
    status: MembershipStatus | None = None

class CompanyContextRead(BaseModel):
    company: CompanyRead
    membership: MembershipRead
    permissions: list[str]
    departments: list[DepartmentRead]


class InvitationCreate(BaseModel):
    email: EmailStr
    member_role: MembershipRole
    department_id: int | None = None


class InvitationRead(BaseModel):
    id: int
    company_id: int
    company_name: str
    email: str
    member_role: MembershipRole
    department_id: int | None = None
    department_name: str | None = None
    status: InvitationStatus
    delivery_status: InvitationDeliveryStatus
    delivery_attempts: int
    message_id: str | None = None
    delivery_error: str | None = None
    sent_at: datetime | None = None
    bounced_at: datetime | None = None
    expires_at: datetime
    created_at: datetime


class InvitationCreated(InvitationRead):
    invite_token: str
    invite_path: str


class InvitationAccept(BaseModel):
    full_name: str | None = Field(default=None, min_length=2, max_length=255)
    password: str | None = Field(default=None, min_length=8, max_length=128)


class JobAssignmentsUpdate(BaseModel):
    membership_ids: list[int] = Field(default_factory=list, max_length=50)


class JobAssignmentRead(BaseModel):
    job_id: int
    membership_ids: list[int]


class JobAssignmentBatchItem(BaseModel):
    job_id: int = Field(gt=0)
    membership_ids: list[int] = Field(default_factory=list, max_length=50)


class JobAssignmentsBatchUpdate(BaseModel):
    assignments: list[JobAssignmentBatchItem] = Field(min_length=1, max_length=200)

    @model_validator(mode="after")
    def validate_unique_jobs(self) -> "JobAssignmentsBatchUpdate":
        job_ids = [item.job_id for item in self.assignments]
        if len(job_ids) != len(set(job_ids)):
            raise ValueError("Mỗi job chỉ được xuất hiện một lần trong batch.")
        return self


class JobAssignmentsBatchRead(BaseModel):
    assignments: list[JobAssignmentRead]


class CompanyActivityRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    actor_user_id: int | None = None
    actor_email: str
    action: str
    target_type: str
    target_id: str | None = None
    target_label: str | None = None
    details_json: dict[str, Any]
    created_at: datetime


class CompanyActivityPage(BaseModel):
    items: list[CompanyActivityRead]
    total: int
    page: int
    page_size: int


class InvitationBounceEvent(BaseModel):
    message_id: str = Field(min_length=3, max_length=255)
    reason: str | None = Field(default=None, max_length=2000)

File: frontend/src/types/company.ts
export type MembershipRole = "hr" | "department_head";
export type MembershipStatus = "active" | "suspended" | "revoked";
export type InvitationStatus = "pending" | "accepted" | "declined" | "revoked" | "expired";
export type InvitationDeliveryStatus = "not_configured" | "pending" | "sent" | "failed" | "bounced";

export interface Company {
  id: number;
  name: string;
  description: string | null;
  logo_url: string | null;
  is_active: boolean;
}

export interface Department {
  id: number;
  company_id: number;
  name: string;
  description: string | null;
  is_active: boolean;
}

export interface CompanyMembership {
  id: number;
  company_id: number;
  user_id: number;
  full_name: string;
  email: string;
  avatar_url: string | null;
  member_role: MembershipRole;
  department_id: number | null;
  department_name: string | null;
  status: MembershipStatus;
  is_owner: boolean;
  joined_at: string;
  updated_at: string;
}

export interface CompanyContext {
  company: Company;
  membership: CompanyMembership;
  permissions: string[];
  departments: Department[];
}

export interface CompanyInvitation {
  id: number;
  company_id: number;
  company_name: string;
  email: string;
  member_role: MembershipRole;
  department_id: number | null;
  department_name: string | null;
  status: InvitationStatus;
  delivery_status: InvitationDeliveryStatus;
  delivery_attempts: number;
  message_id: string | null;
  delivery_error: string | null;
  sent_at: string | null;
  bounced_at: string | null;
  expires_at: string;
  created_at: string;
  invite_token?: string;
  invite_path?: string;
}

export interface JobAssignment {
  job_id: number;
  membership_ids: number[];
}

export interface JobAssignmentsBatch {
  assignments: JobAssignment[];
}

export interface CompanyActivity {
  id: number;
  actor_user_id: number | null;
  actor_email: string;
  action: string;
  target_type: string;
  target_id: string | null;
  target_label: string | null;
  details_json: Record<string, unknown>;
  created_at: string;
}

export interface CompanyActivityPage {
  items: CompanyActivity[];
  total: number;
  page: number;
  page_size: number;
}

export type RecruitmentRequestStatus = "draft" | "submitted" | "approved" | "rejected" | "cancelled";
export type RecruitmentPriority = "low" | "normal" | "high" | "urgent";

export interface RecruitmentRequest {
  id: number;
  company_id: number;
  department_id: number;
  department_name: string;
  requested_by_id: number;
  requester_name: string;
  requester_email: string;
  title: string;
  headcount: number;
  job_type: "full_time" | "part_time" | "internship" | "freelance" | "remote";
  priority: RecruitmentPriority;
  reason: string;
  responsibilities: string;
  requirements: string;
  target_start_date: string | null;
  status: RecruitmentRequestStatus;
  review_note: string | null;
  reviewed_by_id: number | null;
  reviewer_name: string | null;
  submitted_at: string | null;
  reviewed_at: string | null;
  cancelled_at: string | null;
  converted_job_id: number | null;
  converted_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface RecruitmentRequestPage {
  items: RecruitmentRequest[];
  total: number;
  page: number;
  page_size: number;
}

export interface RecruitmentRequestPayload {
  title: string;
  headcount: number;
  job_type: RecruitmentRequest["job_type"];
  priority: RecruitmentPriority;
  reason: string;
  responsibilities: string;
  requirements: string;
  target_start_date?: string;
  submit?: boolean;
}
